### PR TITLE
DEP: Protect some ExcelWriter attributes

### LINKED ISSUE #45572: ENH: Support for Excel features when writing
### Context

Currently `DataFrame.to_excel` supports `freeze_panes`, and there are requests to support other Excel features (#42560). I've gone through the features of openpyxl and Xlsxwriter, here are other Excel features that pandas might support:

- autofilter
- hiding rows/column/worksheets
- password protection
- conditional formatting
- data validation
- dropdown lists
- defined names
- cell comments

Not listed here are styles as I believe the `styler` supports most of these (cc @ahawryluk).

It seems difficult to justify pandas supporting some of these, but not others. On the other hand, implementing all or even just some of them for each engine adds a lot of complexity and maintenance burden to pandas. I haven't checked, but it may also be the case that some features are supported by certain engines but not by others.

While pandas [documents the `writer` of `ExcelWriter` as having no public attributes](https://pandas.pydata.org/docs/reference/api/pandas.ExcelWriter.html), third party engines document using `writer.book` [e.g. xlsxwriter](https://xlsxwriter.readthedocs.io/example_pandas_autofilter.html).

### Proposal 1: Add more arguments to write_excel.

Add other arguments to `to_excel`, such as autofilter, hiding row/columns/worksheets, conditional formatting. At least these three I perceive to be commonly used.

This is the most user friendly option in that it allows users access to these features with a single function call and the ability to switch engines without having to refactor their code. However as mentioned above, this adds complexity to pandas and is a maintenance burden.

### Proposal 2: Make `writer.book` public, prefix all or most other attributes with an `_`.

This would allow users to use the features that the engine supports directly with `ExcelWriter`, e.g.

```
with ExcelWriter("test.xlsx") as writer: 
    df.to_excel(writer)
    worksheet = writer.book.get_worksheet_by_name("Sheet1")
    worksheet...
```

This would require refactoring the `writer.sheets` as in current state it can get out of sync with `writer.book` if e.g. a user does `writer.book.add_sheet(...)`.

This proposal is less complex/maintenance than Proposal 1, more user friendly than Proposal 3.

### Proposal 3: Don't support any Excel features.

Users wanting these features would either be required to (a) write their DataFrames directly with the engine of choice or (b) write their DataFrames with pandas, then open a book with the engine of their choice if the engine supports modifying existing Excel workbooks.

Less complex/maintenance than Proposal 2, but also less user friendly.

I am currently in favor of Proposal 2. I also see a mix of Proposal 1 and Proposal 2 as viable, where some features are supported directly as arguments in `to_excel` with other features require the user to use `writer.book` for others.

cc @pandas-dev/pandas-core

### PR DESCRIPTION
**Pull request by rhshadrach**
- [x] closes #45572 (Replace xxxx with the Github issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

Built on top of #45794

Any attributes we're looking to make protected are deprecated here, allowing users to use them until the next major release. Not deprecating any immutable attributes - users may find these useful and they do not require any extra care.

Docs for ExcelWriter are attached.
[excel_writer.pdf](https://github.com/pandas-dev/pandas/files/7991646/excel_writer.pdf)

### CODE REVIEW

File: doc/source/whatsnew/v1.5.0.rst [hunk 1]
```diff
@@ -183,6 +183,19 @@ use ``series.loc[i:j]``.
...
+:class:`ExcelWriter` attributes
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The attributes and methods of :class:`ExcelWriter` were previously documented as not
+being public. However some third party Excel engines documented accessing
+``ExcelWriter.book`` or ``ExcelWriter.sheets``, and users were utilizing these
+and possibly other attributes. In order to support this, pandas has made some
+attributes and methods public. Other attributes and methods have been deprecated
+and will raise a ``FutureWarning`` when accessed as they will be removed in a future
+version. See the documentation of :class:`ExcelWriter` for further details.
```
Comment (jreback) [inline doc/source/whatsnew/v1.5.0.rst:197]: i guess can list specific attributes that are being depreacted at some point?

File: pandas/io/excel/_xlwt.py [hunk 2]
```diff
@@ -59,25 +59,33 @@ def __init__(
 
         if encoding is None:
             encoding = "ascii"
-        self.book = xlwt.Workbook(encoding=encoding, **engine_kwargs)
-        self.fm_datetime = xlwt.easyxf(num_format_str=self.datetime_format)
-        self.fm_date = xlwt.easyxf(num_format_str=self.date_format)
+        self._book = xlwt.Workbook(encoding=encoding, **engine_kwargs)
+        self.fm_datetime = xlwt.easyxf(num_format_str=self._datetime_format)
```
Comment (jreback) [inline pandas/io/excel/_xlwt.py:63]: these are public?

**Review by jreback — CHANGES_REQUESTED** [review]
looks fine, pls rebase and can merge (followups on docs & so on are fine).

File: doc/source/whatsnew/v1.5.0.rst (hunk 1 above)
Comment (rhshadrach) [reply doc/source/whatsnew/v1.5.0.rst]: I added a list of the public and deprecated ones.

File: pandas/io/excel/_xlwt.py (hunk 2 above)
Comment (rhshadrach) [reply pandas/io/excel/_xlwt.py]: Thanks for catching, they are now deprecated as well.

**Review by jreback — APPROVED** [review]
great merge on green

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @rhshadrach

**Pull request merged**